debug-only: try to reproduce the journalctl --flush issue

mrc0mmand opened the pull request:
Let's try to reproduce the `journalctl --flush` issue described in #12570. @yuwata feel free to push anything you want here, if you're willing to play around with it :-)

evverx commented:
@mrc0mmand judging by https://dev.azure.com/evvers/systemd-systemd/_build/results?buildId=349&view=logs&jobId=56fae37d-e638-5fac-6ab9-ca4fcdfee404&taskId=e3a12e69-71ef-587d-d243-f543ea61fe68&lineStart=8234&lineEnd=8235&colStart=1&colEnd=1, you've just reproduced it :-) Looks like the PR served its purpose :-)

mrc0mmand commented:
That was easier than I thought :-) I wanted to get the debug logs as well and hopefully see what's wrong with it, but it apparently not that obvious:

```
2019-05-24T12:17:07.3224074Z [   52.794474] systemd-journald[201]: systemd-journald running as pid 201
2019-05-24T12:17:07.5435782Z [   53.015512] systemd[1]: systemd-journald.service: Got notification message from PID 201 (READY=1, STATUS=Processing requests...)
2019-05-24T12:17:07.5447755Z [   53.016927] systemd[1]: systemd-journald.service: Changed start -> running
2019-05-24T12:17:07.5455123Z [   53.017635] systemd[1]: systemd-journald.service: Job 41 systemd-journald.service/start finished, result=done
2019-05-24T12:17:07.5456908Z [   53.018001] systemd[1]: Started Journal Service.
2019-05-24T12:17:07.5471109Z [  OK  ] Started Journal Service.
2019-05-24T12:17:07.5482932Z [   53.020465] systemd-journald[201]: Sent READY=1 notification.
2019-05-24T12:17:07.5568074Z [   53.028914] systemd[1]: systemd-journald-dev-log.socket: Changed listening -> running
2019-05-24T12:17:07.6165057Z [   53.088399] audit: type=1130 audit(1558700227.212:11): pid=1 uid=0 auid=4294967295 ses=4294967295 msg='unit=systemd-journald comm="systemd" exe="/lib/systemd/systemd" hostname=? addr=? terminal=? res=success'
2019-05-24T12:17:07.9444316Z          Starting Flush Journal to Persistent Storage...
2019-05-24T12:17:08.0472723Z [   53.519354] systemd-journald[201]: Sent WATCHDOG=1 notification.
2019-05-24T12:17:08.8358258Z [   54.307829] systemd-journald[201]: n/a: New incoming connection.
2019-05-24T12:17:08.8941343Z [   54.366243] systemd-journald[201]: varlink-19: varlink: setting state idle-server
2019-05-24T12:17:08.9164203Z [   54.388476] systemd-journald[201]: varlink-19: New incoming message: {"method":"io.systemd.Journal.FlushToVar","parameters":{}}
2019-05-24T12:17:08.9810964Z [   54.453220] systemd-journald[201]: varlink-19: varlink: changing state idle-server → processing-method
2019-05-24T12:17:08.9821038Z [   54.454255] systemd-journald[201]: Received client request to flush runtime journal.
2019-05-24T12:17:08.9986381Z [   54.470714] systemd-journald[201]: Journal effective settings seal=yes compress=yes compress_threshold_bytes=512B
2019-05-24T12:17:09.0069010Z [   54.478880] systemd-journald[201]: Fixed min_use=1.0M max_use=91.9M max_size=11.4M min_size=512.0K keep_free=137.8M n_max_files=100
2019-05-24T12:17:09.2349262Z [   54.707003] systemd-journald[201]: Reserving 20913 entries in hash table.
2019-05-24T12:17:09.3097985Z [   54.781909] systemd-journald[201]: Flushing to /var...
2019-05-24T12:17:09.3115727Z [   54.783666] systemd-journald[201]: Considering root directory '/run/log/journal'.
2019-05-24T12:17:09.3139094Z [   54.786084] systemd-journald[201]: Root directory /run/log/journal added.
2019-05-24T12:17:09.3155923Z [   54.787647] systemd-journald[201]: Considering directory '/run/log/journal/4173ed685c474cafa70c9ca1b9cb8030'.
2019-05-24T12:17:09.3226677Z [   54.794782] systemd-journald[201]: Directory /run/log/journal/4173ed685c474cafa70c9ca1b9cb8030 added.
2019-05-24T12:17:09.3250996Z [   54.797238] systemd-journald[201]: Journal effective settings seal=no compress=no compress_threshold_bytes=8B
2019-05-24T12:17:09.3330924Z [   54.805195] systemd-journald[201]: File /run/log/journal/4173ed685c474cafa70c9ca1b9cb8030/system.journal added.
2019-05-24T12:17:09.3341430Z [   54.806360] systemd-journald[201]: Considering root directory '/var/log/journal'.
2019-05-24T12:17:09.3345853Z [   54.806745] systemd-journald[201]: Considering root directory '/var/log/journal/remote'.
2019-05-24T12:17:12.7947653Z [    **] (1 of 2) A start job is running for…ersistent Storage (40s / 3min 35s)
2019-05-24T12:17:13.3861319Z [     *] (2 of 2) A start job is running for /dev/ttyS0 (41s / 3min 3s)
2019-05-24T12:17:13.7901453Z [    **] (2 of 2) A start job is running for /dev/ttyS0 (41s / 3min 3s)
2019-05-24T12:17:14.2905313Z [   ***] (2 of 2) A start job is running for /dev/ttyS0 (42s / 3min 3s)
2019-05-24T12:17:14.7905517Z [  *** ] (1 of 2) A start job is running for…ersistent Storage (42s / 3min 35s)
2019-05-24T12:17:15.2908470Z [ ***  ] (1 of 2) A start job is running for…ersistent Storage (43s / 3min 35s)
2019-05-24T12:17:15.8564494Z [***   ] (1 of 2) A start job is running for…ersistent Storage (44s / 3min 35s)
2019-05-24T12:17:16.3044112Z [**    ] (2 of 2) A start job is running for /dev/ttyS0 (44s / 3min 3s)
2019-05-24T12:17:16.7884444Z [*     ] (2 of 2) A start job is running for /dev/ttyS0 (44s / 3min 3s)
2019-05-24T12:17:17.2886158Z [**    ] (2 of 2) A start job is running for /dev/ttyS0 (45s / 3min 3s)
2019-05-24T12:17:17.8363768Z [***   ] (1 of 2) A start job is running for…ersistent Storage (46s / 3min 35s)
2019-05-24T12:17:18.2965369Z [ ***  ] (1 of 2) A start job is running for…ersistent Storage (46s / 3min 35s)
2019-05-24T12:17:18.8044243Z [  *** ] (1 of 2) A start job is running for…ersistent Storage (46s / 3min 35s)
2019-05-24T12:17:19.3044735Z [   ***] (2 of 2) A start job is running for /dev/ttyS0 (47s / 3min 3s)
2019-05-24T12:17:19.7929728Z [    **] (2 of 2) A start job is running for /dev/ttyS0 (47s / 3min 3s)
2019-05-24T12:17:25.4249204Z [     *] (2 of 2) A start job is running for /dev/ttyS0 (53s / 3min 3s)
2019-05-24T12:17:28.8163910Z [    **] (1 of 2) A start job is running for…ersistent Storage (56s / 3min 35s)
2019-05-24T12:17:30.5369073Z [   ***] (1 of 2) A start job is running for…ersistent Storage (58s / 3min 35s)
2019-05-24T12:17:32.6204682Z [  *** ] (1 of 2) A start job is running for…rsistent Storage (1min / 3min 35s)
2019-05-24T12:17:34.1763231Z [ ***  ] (2 of 2) A start job is running for /dev/ttyS0 (1min 2s / 3min 3s)
2019-05-24T12:17:35.4645359Z [***   ] (2 of 2) A start job is running for /dev/ttyS0 (1min 3s / 3min 3s)
2019-05-24T12:17:37.2362744Z [**    ] (2 of 2) A start job is running for /dev/ttyS0 (1min 5s / 3min 3s)
2019-05-24T12:17:39.6194670Z [*     ] (1 of 2) A start job is running for…stent Storage (1min 6s / 3min 35s)
2019-05-24T12:17:41.3188821Z [**    ] (1 of 2) A start job is running for…stent Storage (1min 9s / 3min 35s)
2019-05-24T12:17:43.5526357Z [***   ] (1 of 2) A start job is running for…tent Storage (1min 11s / 3min 35s)
2019-05-24T12:17:45.2284320Z [ ***  ] (2 of 2) A start job is running for /dev/ttyS0 (1min 13s / 3min 3s)
2019-05-24T12:17:47.2044571Z [  *** ] (2 of 2) A start job is running for /dev/ttyS0 (1min 15s / 3min 3s)
2019-05-24T12:17:48.9326816Z [   ***] (2 of 2) A start job is running for /dev/ttyS0 (1min 17s / 3min 3s)
2019-05-24T12:17:50.2368565Z [    **] (1 of 2) A start job is running for…tent Storage (1min 18s / 3min 35s)
2019-05-24T12:17:50.9740517Z [     *] (1 of 2) A start job is running for…tent Storage (1min 19s / 3min 35s)
2019-05-24T12:17:51.6643328Z [    **] (1 of 2) A start job is running for…tent Storage (1min 19s / 3min 35s)
2019-05-24T12:17:52.8364138Z [   ***] (2 of 2) A start job is running for /dev/ttyS0 (1min 21s / 3min 3s)
2019-05-24T12:17:53.5225323Z [  OK  ] Found device /dev/ttyS0.
2019-05-24T12:17:55.0843132Z [FAILED] Failed to start Flush Journal to Persistent Storage.
2019-05-24T12:17:55.0900145Z See 'systemctl status systemd-journal-flush.service' for details.
2019-05-24T12:17:55.1247255Z [  100.596565] audit: type=1130 audit(1558700274.744:12): pid=1 uid=0 auid=4294967295 ses=4294967295 msg='unit=systemd-journal-flush comm="systemd" exe="/lib/systemd/systemd" hostname=? addr=? terminal=? res=failed'
2019-05-24T12:17:55.8164418Z          Starting Create Volatile Files and Directories...
2019-05-24T12:17:55.9166774Z [  OK  ] Listening on Load/Save RF …itch Status /dev/rfkill Watch.

```

evverx commented:
The only way to figure it out I came up with in https://github.com/systemd/systemd/pull/12544 was to also set `DefaultStandardOutput` to `journal+console`. And here is how it fails:
```
[  105.280965] journalctl[229]: n/a: varlink: setting state idle-client
[  105.306710] journalctl[229]: journal: Sending message: {"method":"io.systemd.Journal.FlushToVar","parameters":{}}
[FAILED] Failed to start Flush Journal to Persistent Storage.
See 'systemctl status systemd-journal-flush.service' for details.
[  105.347011] journalctl[229]: journal: varlink: changing state idle-client → calling
[  105.377564] journalctl[229]: journal: varlink: changing state calling → pending-timeout
[  105.406807] journalctl[229]: Failed to execute varlink call: (null)
[  105.597778] journalctl[229]: journal: varlink: changing state pending-timeout → disconnected
```
I'm not sure how to interpret "Failed to execute varlink call: (null)" though :-)

mrc0mmand commented:
Ah, thank you! Added it to the current run for the sake of completeness, if someone would like to debug this further and potentially fix it.

mrc0mmand commented:
This looks kind of scary:

```
==206==ERROR: AddressSanitizer failed to allocate 0x202000 (2105344) bytes of LargeMmapAllocator (errno: 12)
==206==Process memory map follows:
	0x000000400000-0x00000047e000	/lib/systemd/systemd-journald
	0x00000067e000-0x000000680000	/lib/systemd/systemd-journald
	0x000000680000-0x0000006f5000	/lib/systemd/systemd-journald
	0x00007fff7000-0x00008fff7000	
	0x00008fff7000-0x02008fff7000	
	0x02008fff7000-0x10007fff8000	
	0x600000000000-0x602000000000	
	0x602000000000-0x602000040000	
	0x602000040000-0x603000000000	
	0x603000000000-0x603000040000	
	0x603000040000-0x604000000000	
	0x604000000000-0x604000030000	
	0x604000030000-0x606000000000	
	0x606000000000-0x606000020000	
	0x606000020000-0x607000000000	
	0x607000000000-0x607000020000	
	0x607000020000-0x608000000000	
	0x608000000000-0x608000010000	
	0x608000010000-0x60b000000000	
	0x60b000000000-0x60b0000f0000	
	0x60b0000f0000-0x60c000000000	
	0x60c000000000-0x60c000010000	
	0x60c000010000-0x60d000000000	
	0x60d000000000-0x60d000010000	
	0x60d000010000-0x60e000000000	
	0x60e000000000-0x60e000010000	
	0x60e000010000-0x60f000000000	
	0x60f000000000-0x60f000010000	
	0x60f000010000-0x610000000000	
	0x610000000000-0x610000010000	
	0x610000010000-0x611000000000	
	0x611000000000-0x611000010000	
	0x611000010000-0x612000000000	
	0x612000000000-0x612000010000	
	0x612000010000-0x613000000000	
	0x613000000000-0x613000010000	
	0x613000010000-0x614000000000	
	0x614000000000-0x614000020000	
	0x614000020000-0x615000000000	
	0x615000000000-0x615000040000	
	0x615000040000-0x616000000000	
	0x616000000000-0x616000050000	
	0x616000050000-0x617000000000	
	0x617000000000-0x617000020000	
	0x617000020000-0x619000000000	
	0x619000000000-0x619000040000	
	0x619000040000-0x61a000000000	
	0x61a000000000-0x61a000080000	
	0x61a000080000-0x61d000000000	
	0x61d000000000-0x61d000020000	
	0x61d000020000-0x621000000000	
	0x621000000000-0x6210001c0000	
	0x6210001c0000-0x624000000000	
	0x624000000000-0x624000050000	
	0x624000050000-0x625000000000	
	0x625000000000-0x625000020000	
	0x625000020000-0x62b000000000	
	0x62b000000000-0x62b000010000	
	0x62b000010000-0x62d000000000	
	0x62d000000000-0x62d000080000	
	0x62d000080000-0x631000000000	
	0x631000000000-0x631000030000	
	0x631000030000-0x640000000000	
	0x640000000000-0x640000003000	
	0x7f0746fe8000-0x7f07479f2000	
	0x7f07479f2000-0x7f07479f3000	
	0x7f07479f3000-0x7f07481f3000	
	0x7f07481f3000-0x7f07489f3000	/var/log/journal/f4808ae70049452ba8138116258c98ec/system.journal
	0x7f0748bed000-0x7f074a200000	
	0x7f074a300000-0x7f074a400000	
	0x7f074a500000-0x7f074a600000	
	0x7f074a700000-0x7f074a800000	
	0x7f074a8f0000-0x7f074cc42000	
	0x7f074cc42000-0x7f074cc54000	/lib/x86_64-linux-gnu/libgpg-error.so.0.17.0
	0x7f074cc54000-0x7f074ce54000	/lib/x86_64-linux-gnu/libgpg-error.so.0.17.0
	0x7f074ce54000-0x7f074ce55000	/lib/x86_64-linux-gnu/libgpg-error.so.0.17.0
	0x7f074ce55000-0x7f074ce56000	/lib/x86_64-linux-gnu/libgpg-error.so.0.17.0
	0x7f074ce56000-0x7f074ceab000	/lib/x86_64-linux-gnu/libdevmapper.so.1.02.1
	0x7f074ceab000-0x7f074d0aa000	/lib/x86_64-linux-gnu/libdevmapper.so.1.02.1
	0x7f074d0aa000-0x7f074d0ab000	/lib/x86_64-linux-gnu/libdevmapper.so.1.02.1
	0x7f074d0ab000-0x7f074d0ae000	/lib/x86_64-linux-gnu/libdevmapper.so.1.02.1
	0x7f074d0ae000-0x7f074d0af000	
	0x7f074d0af000-0x7f074d0b3000	/lib/x86_64-linux-gnu/libuuid.so.1.3.0
	0x7f074d0b3000-0x7f074d2b2000	/lib/x86_64-linux-gnu/libuuid.so.1.3.0
	0x7f074d2b2000-0x7f074d2b3000	/lib/x86_64-linux-gnu/libuuid.so.1.3.0
	0x7f074d2b3000-0x7f074d2b4000	/lib/x86_64-linux-gnu/libuuid.so.1.3.0
	0x7f074d2b4000-0x7f074d2b8000	/lib/x86_64-linux-gnu/libattr.so.1.1.0
	0x7f074d2b8000-0x7f074d4b7000	/lib/x86_64-linux-gnu/libattr.so.1.1.0
	0x7f074d4b7000-0x7f074d4b8000	/lib/x86_64-linux-gnu/libattr.so.1.1.0
	0x7f074d4b8000-0x7f074d4b9000	/lib/x86_64-linux-gnu/libattr.so.1.1.0
	0x7f074d4b9000-0x7f074d635000	/usr/lib/x86_64-linux-gnu/libstdc++.so.6.0.25
	0x7f074d635000-0x7f074d835000	/usr/lib/x86_64-linux-gnu/libstdc++.so.6.0.25
	0x7f074d835000-0x7f074d83f000	/usr/lib/x86_64-linux-gnu/libstdc++.so.6.0.25
	0x7f074d83f000-0x7f074d841000	/usr/lib/x86_64-linux-gnu/libstdc++.so.6.0.25
	0x7f074d841000-0x7f074d845000	
	0x7f074d845000-0x7f074d8b5000	/lib/x86_64-linux-gnu/libpcre.so.3.13.3
	0x7f074d8b5000-0x7f074dab4000	/lib/x86_64-linux-gnu/libpcre.so.3.13.3
	0x7f074dab4000-0x7f074dab5000	/lib/x86_64-linux-gnu/libpcre.so.3.13.3
	0x7f074dab5000-0x7f074dab6000	/lib/x86_64-linux-gnu/libpcre.so.3.13.3
	0x7f074dab6000-0x7f074daf2000	/lib/x86_64-linux-gnu/libblkid.so.1.1.0
	0x7f074daf2000-0x7f074dcf1000	/lib/x86_64-linux-gnu/libblkid.so.1.1.0
	0x7f074dcf1000-0x7f074dcf5000	/lib/x86_64-linux-gnu/libblkid.so.1.1.0
	0x7f074dcf5000-0x7f074dcf6000	/lib/x86_64-linux-gnu/libblkid.so.1.1.0
	0x7f074dcf6000-0x7f074dcf7000	
	0x7f074dcf7000-0x7f074dd0e000	/usr/lib/x86_64-linux-gnu/liblz4.so.1.7.1
	0x7f074dd0e000-0x7f074df0d000	/usr/lib/x86_64-linux-gnu/liblz4.so.1.7.1
	0x7f074df0d000-0x7f074df0e000	/usr/lib/x86_64-linux-gnu/liblz4.so.1.7.1
	0x7f074df0e000-0x7f074df0f000	/usr/lib/x86_64-linux-gnu/liblz4.so.1.7.1
	0x7f074df0f000-0x7f074df30000	/lib/x86_64-linux-gnu/liblzma.so.5.0.0
	0x7f074df30000-0x7f074e12f000	/lib/x86_64-linux-gnu/liblzma.so.5.0.0
	0x7f074e12f000-0x7f074e130000	/lib/x86_64-linux-gnu/liblzma.so.5.0.0
	0x7f074e130000-0x7f074e131000	/lib/x86_64-linux-gnu/liblzma.so.5.0.0
	0x7f074e131000-0x7f074e162000	/usr/lib/x86_64-linux-gnu/libidn.so.11.6.15
	0x7f074e162000-0x7f074e362000	/usr/lib/x86_64-linux-gnu/libidn.so.11.6.15
	0x7f074e362000-0x7f074e363000	/usr/lib/x86_64-linux-gnu/libidn.so.11.6.15
	0x7f074e363000-0x7f074e364000	/usr/lib/x86_64-linux-gnu/libidn.so.11.6.15
	0x7f074e364000-0x7f074e392000	/lib/x86_64-linux-gnu/libseccomp.so.2.3.1
	0x7f074e392000-0x7f074e592000	/lib/x86_64-linux-gnu/libseccomp.so.2.3.1
	0x7f074e592000-0x7f074e5a8000	/lib/x86_64-linux-gnu/libseccomp.so.2.3.1
	0x7f074e5a8000-0x7f074e5a9000	/lib/x86_64-linux-gnu/libseccomp.so.2.3.1
	0x7f074e5a9000-0x7f074e5ed000	/lib/x86_64-linux-gnu/libmount.so.1.1.0
	0x7f074e5ed000-0x7f074e7ed000	/lib/x86_64-linux-gnu/libmount.so.1.1.0
	0x7f074e7ed000-0x7f074e7ee000	/lib/x86_64-linux-gnu/libmount.so.1.1.0
	0x7f074e7ee000-0x7f074e7ef000	/lib/x86_64-linux-gnu/libmount.so.1.1.0
	0x7f074e7ef000-0x7f074e7f0000	
	0x7f074e7f0000-0x7f074e806000	/lib/x86_64-linux-gnu/libkmod.so.2.3.0
	0x7f074e806000-0x7f074ea05000	/lib/x86_64-linux-gnu/libkmod.so.2.3.0
	0x7f074ea05000-0x7f074ea06000	/lib/x86_64-linux-gnu/libkmod.so.2.3.0
	0x7f074ea06000-0x7f074ea07000	/lib/x86_64-linux-gnu/libkmod.so.2.3.0
	0x7f074ea07000-0x7f074ea0d000	/lib/x86_64-linux-gnu/libip4tc.so.0.1.0
	0x7f074ea0d000-0x7f074ec0c000	/lib/x86_64-linux-gnu/libip4tc.so.0.1.0
	0x7f074ec0c000-0x7f074ec0d000	/lib/x86_64-linux-gnu/libip4tc.so.0.1.0
	0x7f074ec0d000-0x7f074ec0e000	/lib/x86_64-linux-gnu/libip4tc.so.0.1.0
	0x7f074ec0e000-0x7f074ece5000	/lib/x86_64-linux-gnu/libgcrypt.so.20.0.5
	0x7f074ece5000-0x7f074eee5000	/lib/x86_64-linux-gnu/libgcrypt.so.20.0.5
	0x7f074eee5000-0x7f074eee6000	/lib/x86_64-linux-gnu/libgcrypt.so.20.0.5
	0x7f074eee6000-0x7f074eeee000	/lib/x86_64-linux-gnu/libgcrypt.so.20.0.5
	0x7f074eeee000-0x7f074eeef000	
	0x7f074eeef000-0x7f074ef14000	/lib/x86_64-linux-gnu/libcryptsetup.so.4.6.0
	0x7f074ef14000-0x7f074f113000	/lib/x86_64-linux-gnu/libcryptsetup.so.4.6.0
	0x7f074f113000-0x7f074f114000	/lib/x86_64-linux-gnu/libcryptsetup.so.4.6.0
	0x7f074f114000-0x7f074f116000	/lib/x86_64-linux-gnu/libcryptsetup.so.4.6.0
	0x7f074f116000-0x7f074f11d000	/lib/x86_64-linux-gnu/libacl.so.1.1.0
	0x7f074f11d000-0x7f074f31c000	/lib/x86_64-linux-gnu/libacl.so.1.1.0
	0x7f074f31c000-0x7f074f31d000	/lib/x86_64-linux-gnu/libacl.so.1.1.0
	0x7f074f31d000-0x7f074f31e000	/lib/x86_64-linux-gnu/libacl.so.1.1.0
	0x7f074f31e000-0x7f074f322000	/lib/x86_64-linux-gnu/libcap.so.2.24
	0x7f074f322000-0x7f074f522000	/lib/x86_64-linux-gnu/libcap.so.2.24
	0x7f074f522000-0x7f074f523000	/lib/x86_64-linux-gnu/libcap.so.2.24
	0x7f074f523000-0x7f074f524000	/lib/x86_64-linux-gnu/libcap.so.2.24
	0x7f074f524000-0x7f074f52b000	/lib/x86_64-linux-gnu/librt-2.23.so
	0x7f074f52b000-0x7f074f72a000	/lib/x86_64-linux-gnu/librt-2.23.so
	0x7f074f72a000-0x7f074f72b000	/lib/x86_64-linux-gnu/librt-2.23.so
	0x7f074f72b000-0x7f074f72c000	/lib/x86_64-linux-gnu/librt-2.23.so
	0x7f074f72c000-0x7f074f743000	/lib/x86_64-linux-gnu/libgcc_s.so.1
	0x7f074f743000-0x7f074f942000	/lib/x86_64-linux-gnu/libgcc_s.so.1
	0x7f074f942000-0x7f074f943000	/lib/x86_64-linux-gnu/libgcc_s.so.1
	0x7f074f943000-0x7f074f944000	/lib/x86_64-linux-gnu/libgcc_s.so.1
	0x7f074f944000-0x7f074fa4c000	/lib/x86_64-linux-gnu/libm-2.23.so
	0x7f074fa4c000-0x7f074fc4b000	/lib/x86_64-linux-gnu/libm-2.23.so
	0x7f074fc4b000-0x7f074fc4c000	/lib/x86_64-linux-gnu/libm-2.23.so
	0x7f074fc4c000-0x7f074fc4d000	/lib/x86_64-linux-gnu/libm-2.23.so
	0x7f074fc4d000-0x7f074fc50000	/lib/x86_64-linux-gnu/libdl-2.23.so
	0x7f074fc50000-0x7f074fe4f000	/lib/x86_64-linux-gnu/libdl-2.23.so
	0x7f074fe4f000-0x7f074fe50000	/lib/x86_64-linux-gnu/libdl-2.23.so
	0x7f074fe50000-0x7f074fe51000	/lib/x86_64-linux-gnu/libdl-2.23.so
	0x7f074fe51000-0x7f0750011000	/lib/x86_64-linux-gnu/libc-2.23.so
	0x7f0750011000-0x7f0750211000	/lib/x86_64-linux-gnu/libc-2.23.so
	0x7f0750211000-0x7f0750215000	/lib/x86_64-linux-gnu/libc-2.23.so
	0x7f0750215000-0x7f0750217000	/lib/x86_64-linux-gnu/libc-2.23.so
	0x7f0750217000-0x7f075021b000	
	0x7f075021b000-0x7f0750233000	/lib/x86_64-linux-gnu/libpthread-2.23.so
	0x7f0750233000-0x7f0750432000	/lib/x86_64-linux-gnu/libpthread-2.23.so
	0x7f0750432000-0x7f0750433000	/lib/x86_64-linux-gnu/libpthread-2.23.so
	0x7f0750433000-0x7f0750434000	/lib/x86_64-linux-gnu/libpthread-2.23.so
	0x7f0750434000-0x7f0750438000	
	0x7f0750438000-0x7f0750485000	/usr/lib/x86_64-linux-gnu/libubsan.so.0.0.0
	0x7f0750485000-0x7f0750684000	/usr/lib/x86_64-linux-gnu/libubsan.so.0.0.0
	0x7f0750684000-0x7f0750686000	/usr/lib/x86_64-linux-gnu/libubsan.so.0.0.0
	0x7f0750686000-0x7f0750689000	/usr/lib/x86_64-linux-gnu/libubsan.so.0.0.0
	0x7f0750689000-0x7f075115a000	
	0x7f075115a000-0x7f0751179000	/lib/x86_64-linux-gnu/libselinux.so.1
	0x7f0751179000-0x7f0751378000	/lib/x86_64-linux-gnu/libselinux.so.1
	0x7f0751378000-0x7f0751379000	/lib/x86_64-linux-gnu/libselinux.so.1
	0x7f0751379000-0x7f075137a000	/lib/x86_64-linux-gnu/libselinux.so.1
	0x7f075137a000-0x7f075137c000	
	0x7f075137c000-0x7f0751f2a000	/lib/systemd/libsystemd-shared-242.so
	0x7f0751f2a000-0x7f075212a000	/lib/systemd/libsystemd-shared-242.so
	0x7f075212a000-0x7f07521bb000	/lib/systemd/libsystemd-shared-242.so
	0x7f07521bb000-0x7f0752761000	/lib/systemd/libsystemd-shared-242.so
	0x7f0752761000-0x7f0752762000	
	0x7f0752762000-0x7f0752856000	/usr/lib/x86_64-linux-gnu/libasan.so.2.0.0
	0x7f0752856000-0x7f0752a56000	/usr/lib/x86_64-linux-gnu/libasan.so.2.0.0
	0x7f0752a56000-0x7f0752a59000	/usr/lib/x86_64-linux-gnu/libasan.so.2.0.0
	0x7f0752a59000-0x7f0752a5a000	/usr/lib/x86_64-linux-gnu/libasan.so.2.0.0
	0x7f0752a5a000-0x7f07536cf000	
	0x7f07536cf000-0x7f07536f5000	/lib/x86_64-linux-gnu/ld-2.23.so
	0x7f075376b000-0x7f0753783000	
	0x7f0753783000-0x7f0753784000	/var/log/journal/f4808ae70049452ba8138116258c98ec/system.journal
	0x7f0753784000-0x7f07537a6000	
	0x7f07537a7000-0x7f07537a8000	/run/systemd/journal/kernel-seqnum
	0x7f07537a8000-0x7f07538c6000	
	0x7f07538c6000-0x7f07538e4000	/lib/x86_64-linux-gnu/libudev.so.1.6.4
	0x7f07538e4000-0x7f07538e5000	/lib/x86_64-linux-gnu/libudev.so.1.6.4
	0x7f07538e5000-0x7f07538e6000	/lib/x86_64-linux-gnu/libudev.so.1.6.4
	0x7f07538e6000-0x7f07538f1000	
	0x7f07538f1000-0x7f07538f4000	
	0x7f07538f4000-0x7f07538f5000	/lib/x86_64-linux-gnu/ld-2.23.so
	0x7f07538f5000-0x7f07538f6000	/lib/x86_64-linux-gnu/ld-2.23.so
	0x7f07538f6000-0x7f07538f7000	
	0x7ffc75982000-0x7ffc759a3000	[stack]
	0x7ffc759cd000-0x7ffc759d0000	[vvar]
	0x7ffc759d0000-0x7ffc759d2000	[vdso]
	0xffffffffff600000-0xffffffffff601000	[vsyscall]
==206==End of process memory map.
==206==AddressSanitizer CHECK failed: ../../../../src/libsanitizer/sanitizer_common/sanitizer_posix.cc:121 "(("unable to mmap" && 0)) != (0)" (0x0, 0x0)
    #0 0x7f0752802631  (/usr/lib/x86_64-linux-gnu/libasan.so.2.0.0+0xa0631)
    #1 0x7f07528075e3 in __sanitizer::CheckFailed(char const*, int, char const*, unsigned long long, unsigned long long) (/usr/lib/x86_64-linux-gnu/libasan.so.2.0.0+0xa55e3)
    #2 0x7f075280f611  (/usr/lib/x86_64-linux-gnu/libasan.so.2.0.0+0xad611)
    #3 0x7f0752784c0c  (/usr/lib/x86_64-linux-gnu/libasan.so.2.0.0+0x22c0c)
    #4 0x7f07527fa5d2 in malloc (/usr/lib/x86_64-linux-gnu/libasan.so.2.0.0+0x985d2)
    #5 0x7f0751ae51a6 in malloc_multiply ../src/basic/alloc-util.h:96
    #6 0x7f0751ae6474 in get_process_cmdline ../src/basic/process-util.c:145
    #7 0x44d229 in client_context_read_basic ../src/journal/journald-context.c:233
    #8 0x4516b8 in client_context_really_refresh ../src/journal/journald-context.c:514
    #9 0x45205b in client_context_maybe_refresh ../src/journal/journald-context.c:577
    #10 0x452f90 in client_context_get_internal ../src/journal/journald-context.c:676
    #11 0x453336 in client_context_get ../src/journal/journald-context.c:714
    #12 0x4574b7 in server_process_native_message ../src/journal/journald-native.c:310
    #13 0x426012 in server_process_datagram ../src/journal/journald-server.c:1346
    #14 0x7f0751cbfad3 in source_dispatch ../src/libsystemd/sd-event/sd-event.c:2827
    #15 0x7f0751cc45b4 in sd_event_dispatch ../src/libsystemd/sd-event/sd-event.c:3240
    #16 0x7f0751cc5254 in sd_event_run ../src/libsystemd/sd-event/sd-event.c:3297
    #17 0x40737d in main ../src/journal/journald.c:92
    #18 0x7f074fe7182f in __libc_start_main (/lib/x86_64-linux-gnu/libc.so.6+0x2082f)
    #19 0x406cf8 in _start (/lib/systemd/systemd-journald+0x406cf8)
```

Is this just a fancy way of saying the VM ran out of memory? :-)

evverx commented:
Let's start with the good news :-) As far as I can tell, the test was run 3 times and it seems that `journalctl --flush` didn't fail, which seems to suggest that setting the timeout to 120 seconds somewhat helps.

Regarding the ASan report, it says it crashed because it couldn't allocate enough memory and, in theory, these errors could be masked by setting [`allocator_may_return_null`](https://github.com/google/sanitizers/wiki/SanitizerCommonFlags) to `true`. Though it will most likely lead to cascading failure anyway.

I suspect `QEMU_MEM` and `QEMU_SMP` aren't suitable to be used on Azure Pipelines and QEMU_MEM could be bumped up temporarily to keep experimenting with varlink. But in general I think it would make sense to run `lscpu` on CentOS CI in VMs where Arch is run and also on Azure Pipelines to figure out what we're dealing with :-) It could be that VMs run with Vagrant have only one core, which, in turn, limits the number of processes competing for memory at the same time.

mrc0mmand commented:
> I suspect `QEMU_MEM` and `QEMU_SMP` aren't suitable to be used on Azure Pipelines and QEMU_MEM could be bumped up temporarily to keep experimenting with varlink. But in general I think it would make sense to run `lscpu` on CentOS CI in VMs where Arch is run and also on Azure Pipelines to figure out what we're dealing with :-) It could be that VMs run with Vagrant have only one core, which, in turn, limits the number of processes competing for memory at the same time.

In CentOS CI each baremetal machine has 8 cores thus I run each Vagrant VM with [8 VCPUs](https://github.com/systemd/systemd-centos-ci/blob/master/vagrant/Vagrantfiles/Vagrantfile_arch-sanitizers-clang#L58). Also, for each TEST-??-* I set the `QEMU_SMP` [to the value of](https://github.com/systemd/systemd-centos-ci/blob/master/vagrant/vagrant-test-sanitizers.sh#L29) `nproc`, so it could benefit from all resources the machine provides. For *standard* runs it's more complicated due to parallelism, but that shouldn't matter in this case.

Maybe we could tweak the `QEMU_SMP` in Azure Pipelines as well, will try to get the output of `lscpu` or `/proc/cpuinfo`.

mrc0mmand commented:
Azure Pipelines:

```
+ cat /proc/cpuinfo
processor	: 0
vendor_id	: GenuineIntel
cpu family	: 6
model		: 63
model name	: Intel(R) Xeon(R) CPU E5-2673 v3 @ 2.40GHz
stepping	: 2
microcode	: 0xffffffff
cpu MHz		: 2394.455
cache size	: 30720 KB
physical id	: 0
siblings	: 2
core id		: 0
cpu cores	: 2
apicid		: 0
initial apicid	: 0
fpu		: yes
fpu_exception	: yes
cpuid level	: 15
wp		: yes
flags		: fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush mmx fxsr sse sse2 ss ht syscall nx pdpe1gb rdtscp lm constant_tsc rep_good nopl xtopology cpuid pni pclmulqdq ssse3 fma cx16 pcid sse4_1 sse4_2 movbe popcnt aes xsave avx f16c rdrand hypervisor lahf_lm abm invpcid_single pti fsgsbase bmi1 avx2 smep bmi2 erms invpcid xsaveopt
bugs		: cpu_meltdown spectre_v1 spectre_v2 spec_store_bypass l1tf
bogomips	: 4788.91
clflush size	: 64
cache_alignment	: 64
address sizes	: 44 bits physical, 48 bits virtual
power management:

processor	: 1
vendor_id	: GenuineIntel
cpu family	: 6
model		: 63
model name	: Intel(R) Xeon(R) CPU E5-2673 v3 @ 2.40GHz
stepping	: 2
microcode	: 0xffffffff
cpu MHz		: 2394.455
cache size	: 30720 KB
physical id	: 0
siblings	: 2
core id		: 1
cpu cores	: 2
apicid		: 1
initial apicid	: 1
fpu		: yes
fpu_exception	: yes
cpuid level	: 15
wp		: yes
flags		: fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush mmx fxsr sse sse2 ss ht syscall nx pdpe1gb rdtscp lm constant_tsc rep_good nopl xtopology cpuid pni pclmulqdq ssse3 fma cx16 pcid sse4_1 sse4_2 movbe popcnt aes xsave avx f16c rdrand hypervisor lahf_lm abm invpcid_single pti fsgsbase bmi1 avx2 smep bmi2 erms invpcid xsaveopt
bugs		: cpu_meltdown spectre_v1 spectre_v2 spec_store_bypass l1tf
bogomips	: 4788.91
clflush size	: 64
cache_alignment	: 64
address sizes	: 44 bits physical, 48 bits virtual
power management:

++ nproc
+ export QEMU_SMP=2
+ QEMU_SMP=2
+ export QEMU_MEM=2048
+ QEMU_MEM=2048
```

Nevertheless, it's still not enough:

```
[  147.285744] CPU: 0 PID: 1 Comm: systemd Not tainted 4.15.0-1041-azure #45-Ubuntu
[  147.285945] Hardware name: QEMU Standard PC (i440FX + PIIX, 1996), BIOS Ubuntu-1.8.2-1ubuntu1 04/01/2014
[  147.286279] Call Trace:
[  147.287095]  dump_stack+0x63/0x82
[  147.287241]  dump_header+0x77/0x285
[  147.287361]  ? security_capable_noaudit+0x4b/0x70
[  147.287489]  oom_kill_process+0x22e/0x450
[  147.287593]  out_of_memory+0x11d/0x4c0
[  147.287733]  __alloc_pages_slowpath+0xd88/0xe70
[  147.287882]  __alloc_pages_nodemask+0x265/0x280
[  147.288052]  alloc_pages_current+0x6a/0xe0
[  147.288158]  __page_cache_alloc+0x86/0x90
[  147.288167]  filemap_fault+0x340/0x750
[  147.288167]  ? filemap_map_pages+0x185/0x3f0
[  147.288167]  ext4_filemap_fault+0x31/0x44
[  147.288167]  __do_fault+0x24/0xec
[  147.288167]  __handle_mm_fault+0xd05/0x11b0
[  147.288167]  ? __switch_to_asm+0x34/0x70
[  147.288167]  ? __switch_to_asm+0x40/0x70
[  147.288167]  handle_mm_fault+0xcc/0x1c0
[  147.288167]  __do_page_fault+0x260/0x500
[  147.288167]  do_page_fault+0x38/0x130
[  147.288167]  ? page_fault+0x2f/0x50
[  147.288167]  page_fault+0x45/0x50
[  147.288167] RIP: 0033:0x50be5b
[  147.288167] RSP: 002b:00007ffc9c8c7058 EFLAGS: 00000246
[  147.288167] RAX: 000061100000db00 RBX: 00000fff93918e14 RCX: 00000000089f322c
[  147.288167] RDX: 000061b00001f180 RSI: 00000000089f322c RDI: 000061100000db00
[  147.288167] RBP: 00007ffc9c8c7170 R08: 0000000000007000 R09: 0000000000001000
[  147.288167] R10: 0000000000002000 R11: 0000000000000246 R12: 00007ffc9c8c70a0
[  147.288167] R13: 000000000050be5b R14: 000061b00001f180 R15: 00007ffc9c8c71a0
[  147.308255] Mem-Info:
[  147.308601] active_anon:352656 inactive_anon:58 isolated_anon:0
[  147.308601]  active_file:29 inactive_file:24 isolated_file:0
[  147.308601]  unevictable:0 dirty:0 writeback:0 unstable:0
[  147.308601]  slab_reclaimable:3477 slab_unreclaimable:4584
[  147.308601]  mapped:2 shmem:65 pagetables:2449 bounce:0
[  147.308601]  free:12722 free_pcp:0 free_cma:0
[  147.309868] Node 0 active_anon:1410624kB inactive_anon:232kB active_file:116kB inactive_file:96kB unevictable:0kB isolated(anon):0kB isolated(file):0kB mapped:8kB dirty:0kB writeback:0kB shmem:260kB shmem_thp: 0kB shmem_pmdmapped: 0kB anon_thp: 1243136kB writeback_tmp:0kB unstable:0kB all_unreclaimable? no
[  147.310732] Node 0 DMA free:6332kB min:468kB low:584kB high:700kB active_anon:9124kB inactive_anon:0kB active_file:0kB inactive_file:0kB unevictable:0kB writepending:0kB present:15992kB managed:15908kB mlocked:0kB kernel_stack:0kB pagetables:72kB bounce:0kB free_pcp:0kB local_pcp:0kB free_cma:0kB
[  147.311431] lowmem_reserve[]: 0 1470 1470 1470 1470
[  147.311585] Node 0 DMA32 free:44556kB min:44584kB low:55728kB high:66872kB active_anon:1401500kB inactive_anon:232kB active_file:116kB inactive_file:100kB unevictable:0kB writepending:0kB present:1556348kB managed:1509600kB mlocked:0kB kernel_stack:1696kB pagetables:9724kB bounce:0kB free_pcp:0kB local_pcp:0kB free_cma:0kB
[  147.312205] lowmem_reserve[]: 0 0 0 0 0
[  147.312205] Node 0 DMA: 23*4kB (UME) 18*8kB (UME) 5*16kB (UM) 2*32kB (U) 3*64kB (UME) 3*128kB (UME) 3*256kB (UME) 3*512kB (UME) 1*1024kB (E) 1*2048kB (E) 0*4096kB = 6332kB
[  147.320657] Node 0 DMA32: 361*4kB (UME) 81*8kB (UME) 94*16kB (UME) 74*32kB (UME) 187*64kB (UME) 64*128kB (UME) 18*256kB (UME) 15*512kB (UME) 6*1024kB (UME) 0*2048kB 0*4096kB = 44556kB
[  147.321218] Node 0 hugepages_total=0 hugepages_free=0 hugepages_surp=0 hugepages_size=2048kB
[  147.321440] 119 total pagecache pages
[  147.321542] 0 pages in swap cache
[  147.321630] Swap cache stats: add 0, delete 0, find 0/0
[  147.321796] Free swap  = 0kB
[  147.321880] Total swap = 0kB
[  147.322045] 393085 pages RAM
[  147.322137] 0 pages HighMem/MovableOnly
[  147.322230] 11708 pages reserved
[  147.322311] 0 pages cma reserved
[  147.322393] 0 pages hwpoisoned
[  147.322511] [ pid ]   uid  tgid total_vm      rss pgtables_bytes swapents oom_score_adj name
[  147.322844] [  209]     0   209 5368766433    37837  1101824        0             0 systemd-journal
[  147.323102] [  221]     0   221 5368751765    38744  1114112        0         -1000 systemd-udevd
[  147.323291] [  252]     0   252 5368751781    38553  1089536        0             0 systemd-udevd
[  147.323493] [  253]     0   253 5368751775    38001  1081344        0             0 systemd-udevd
[  147.323721] [  255]     0   255 5368751775    38011  1081344        0             0 systemd-udevd
[  147.323951] [  256]     0   256 5368751775    38634  1089536        0             0 systemd-udevd
[  147.324151] [  259]     0   259 5368751775    38587  1089536        0             0 systemd-udevd
[  147.324208] [  260]     0   260 5368751775    37994  1081344        0             0 systemd-udevd
[  147.324208] [  264]     0   264 5368751775    38704  1089536        0             0 systemd-udevd
[  147.324208] [  265]     0   265 5368751775    38680  1089536        0             0 systemd-udevd
[  147.324208] [  267]     0   267 5368749540    25311   823296        0             0 systemd-logind
[  147.324208] [  271]     0   271 5368728241    13683   503808        0             0 agetty
[  147.324208] [  272]     0   272 5368751775    38614  1089536        0             0 systemd-udevd
[  147.324208] [  273]   107   273 5368734666    24888   897024        0          -900 dbus-daemon
[  147.324208] [  276]     0   276 5368749738    16563   724992        0             0 systemd-update-
[  147.324208] [  277]     0   277 5368726157     9376   344064        0             0 sh
[  147.324208] [  280]     0   280    27624      695   188416        0             0 systemctl
[  147.324208] Out of memory: Kill process 264 (systemd-udevd) score 99 or sacrifice child
[  147.324208] Killed process 264 (systemd-udevd) total-vm:21475007100kB, anon-rss:154816kB, file-rss:0kB, shmem-rss:0kB
[  147.484368] oom_reaper: reaped process 264 (systemd-udevd), now anon-rss:0kB, file-rss:0kB, shmem-rss:0kB
```

or

```
==201==ERROR: AddressSanitizer failed to allocate 0x202000 (2105344) bytes of LargeMmapAllocator (errno: 12)
ERROR: Failed to mmap
```

mrc0mmand commented:
Ah, scratch that, my `QEMU_MEM` bump was getting overwritten in ASan-specific code block in test/test-functions, let's try it again :-)

evverx commented:
> QEMU_MEM bump was getting overwritten in ASan-specific code block in test/test-functions

On the one hand, that wasn't cool :-) On the other hand, that's a perfect example of how load-bearing values can be hard-coded so reliably that nobody can override them :-) I have to admit it wrote that code and it looked like a good idea at the time but now it's kind of overly defensive and should probably be changed.

Regarding CentOS CI, today I learned about it much more than I expected :-) Anyway, my first hypothesis can be safely thrown away and another one would be that `KERNEL_APPEND=debug` is a stressful experience in terms of resources (and especially for `journald`, which receives all that gibberish). It doesn't kick in on CentOS CI because `KERNEL_APPEND=debug` isn't used there.

mrc0mmand commented:
Given the ASan job in Azure Pipelines is now has been successfully running over and over for past 45 minutes, I'll try to slowly un-workaround it (i.e. drop the `QEMU_MEM` bump and `KERNEL_APPEND=debug`) to see how effective the @yuwata's patches really are.

evverx commented:
Given that we seem to hit this every time we try to debug anything by setting log_level to debug (https://github.com/systemd/systemd/pull/12544/commits/da084c8d5a1907316019cc2558c3dc4b00729e85) I think it'd be better to use 2048 permanently so as not to run into it over and over again. Plus the machine I was talking about in https://github.com/systemd/systemd/pull/11393#issuecomment-453736762 was turned off a long time ago :-)

evverx commented:
And just to clarify, in case my comment was ambiguous (which happens sometimes), when I said "that wasn't cool" I meant that the code resetting QEMU_MEM shouldn't have done that. I didn't mean to say that it's unreasonable to expect that to work or to say that it's weird (it absolutely is) or anything like that.

mrc0mmand commented:
> Given that we seem to hit this every time we try to debug anything by setting log_level to debug ([da084c8](https://github.com/systemd/systemd/commit/da084c8d5a1907316019cc2558c3dc4b00729e85)) I think it'd be better to use 2048 permanently so as not to run into it over and over again.

That would be great.

> Plus the machine I was talking about in [#11393 (comment)](https://github.com/systemd/systemd/pull/11393#issuecomment-453736762) was turned off a long time ago :-)

Ah, I was wondering where this somewhat specific constant came from and completely forgot this discussion :-)

> And just to clarify, in case my comment was ambiguous (which happens sometimes), when I said "that wasn't cool" I meant that the code resetting QEMU_MEM shouldn't have done that. I didn't mean to say that it's unreasonable to expect that to work or to say that it's weird (it absolutely is) or anything like that.

In this case I understood it as you meant it. Maybe in this case it would make sense to enforce a minimum value for `QEMU_MEM`, but the current solution is reasonably safe without any additional hassle.

mrc0mmand commented:
It looks pretty stable!

```
+ echo 'Iteration #13'
+ make -C test/TEST-01-BASIC clean setup run QEMU_TIMEOUT=900 TEST_NO_NSPAWN=yes
Iteration #13
+ (( i++ ))
+ (( i < 30 ))
+ echo 'Iteration #14'
+ make -C test/TEST-01-BASIC clean setup run QEMU_TIMEOUT=900 TEST_NO_NSPAWN=yes
Iteration #14
##[error]The operation was canceled.
##[section]Finishing: This is where it gets darker
#[Error 1]
The job running on agent Hosted Ubuntu 1604 2 has exceeded the maximum execution time of 60. For more information, see https://go.microsoft.com/fwlink/?linkid=2077134 
```

evverx commented on src/journal/journalctl.c:
```diff
@@ -1952,10 +1952,14 @@ static int simple_varlink_call(const char *option, const char *method) {
                 return log_error_errno(r, "Failed to connect to /run/systemd/journal/io.systemd.journal: %m");
 
         (void) varlink_set_description(link, "journal");
+        (void) varlink_set_relative_timeout(link, 120 * USEC_PER_SEC);
```
Now that we know it seems to be working I'm wondering what should happen to the timeout set in the unit file: 0775b9b61120f8be99c and 75f709fbf284041. Before varlink was merged it had been relatively easy to tweak that timeout (in asan_wrapper for example) to adapt the service for almost any environment. I'm not sure I understand how all these timeouts are supposed to interact with each other.

evverx replied:
By the way, I haven't wrapped my head around varlink yet so it's absolutely possible that I'm missing something :-)

yuwata commented:
@mrc0mmand and @evverx I've updated my commits in this PR. Now, the timeout is configurable and unit file disables the varlink timeout. PTAL.

yuwata commented:
So, now the effective timeout is 90s, it may be necessary to bump the value through the unit file.

mrc0mmand commented:
Hmm, I'm not sure if I'm not missing something, but this shouldn't happen with `JOURNALCTL_TIMEOUT=infinity`, right?

```
         Starting ‌Flush Journal to Persistent Storage‌...‌
[   60.063667] systemd-journald[206]: Successfully sent stream file descriptor to service manager.
[   60.187773] systemd-journald[206]: Successfully sent stream file descriptor to service manager.
[   60.278795] systemd-journald[206]: Successfully sent stream file descriptor to service manager.
[   60.337228] systemd-journald[206]: Successfully sent stream file descriptor to service manager.
[   60.575367] systemd-journald[206]: Successfully sent stream file descriptor to service manager.
[   60.704693] systemd-journald[206]: Successfully sent stream file descriptor to service manager.
[   60.783390] systemd-journald[206]: Successfully sent stream file descriptor to service manager.
[   60.899988] systemd-journald[206]: Successfully sent stream file descriptor to service manager.
[   61.032755] systemd-journald[206]: n/a: New incoming connection.
[   61.035049] systemd-journald[206]: varlink-20: varlink: setting state idle-server
[   61.100007] systemd-journald[206]: varlink-20: New incoming message: {"method":"io.systemd.Journal.FlushToVar","parameters":{}}
[   61.169674] systemd-journald[206]: varlink-20: varlink: changing state idle-server → processing-method
[   61.170740] systemd-journald[206]: Received client request to flush runtime journal.
[   61.179164] systemd-journald[206]: Journal effective settings seal=yes compress=yes compress_threshold_bytes=512B
[   61.187593] systemd-journald[206]: Fixed min_use=1.0M max_use=188.7M max_size=23.5M min_size=512.0K keep_free=283.1M n_max_files=100
[   61.446303] systemd-journald[206]: Reserving 42951 entries in hash table.
[   61.485974] systemd-journald[206]: Flushing to /var...
[   61.489058] systemd-journald[206]: Considering root directory '/run/log/journal'.
[   61.491072] systemd-journald[206]: Root directory /run/log/journal added.
[   61.493323] systemd-journald[206]: Considering directory '/run/log/journal/e1a1e13bf3f947c287683a8b7f492e8f'.
[   61.500732] systemd-journald[206]: Directory /run/log/journal/e1a1e13bf3f947c287683a8b7f492e8f added.
[   61.502774] systemd-journald[206]: Journal effective settings seal=no compress=no compress_threshold_bytes=8B
[   61.510675] systemd-journald[206]: File /run/log/journal/e1a1e13bf3f947c287683a8b7f492e8f/system.journal added.
[   61.516400] systemd-journald[206]: Considering root directory '/var/log/journal'.
[   61.516957] systemd-journald[206]: Considering root directory '/var/log/journal/remote'.
[   97.560305] hrtimer: interrupt took 24086360 ns
[‌  OK  ‌] Found device ‌&#x2F;dev&#x2F;ttyS0‌.‌
[‌FAILED‌] Failed to start ‌Flush Journal to Persistent Storage‌.‌
See 'systemctl status systemd-journal-flush.service' for details.
[  107.241860] audit: type=1130 audit(1558991480.252:10): pid=1 uid=0 auid=4294967295 ses=4294967295 msg='unit=systemd-journal-flush comm="systemd" exe="/lib/systemd/systemd" hostname=? addr=? terminal=? res=failed'

...

[  111.367284] journalctl[235]: ‌5mn&#x2F;a: varlink: setting state idle-client‌
[  111.403729] systemd-journald[206]: Journal effective settings seal=no compress=yes compress_threshold_bytes=512B
[  111.404559] systemd[1]: systemd-journald.service: Received EPOLLHUP on stored fd 47 (stored), closing.
[  111.425401] journalctl[235]: ‌5mjournal: Sending message: {&quot;method&quot;:&quot;io.systemd.Journal.FlushToVar&quot;,&quot;parameters&quot;:{}}‌
[  111.427761] systemd-journald[206]: Journal effective settings seal=no compress=yes compress_threshold_bytes=512B
[  111.437085] systemd[1]: systemd-journald.service: Received EPOLLHUP on stored fd 48 (stored), closing.
[  111.439921] journalctl[235]: ‌5mjournal: varlink: changing state idle-client → calling‌
[  111.442292] systemd-journald[206]: Journal effective settings seal=no compress=yes compress_threshold_bytes=512B
[  111.450902] journalctl[235]: ‌5mjournal: varlink: changing state calling → pending-timeout‌
[  111.453240] systemd-journald[206]: Journal effective settings seal=no compress=yes compress_threshold_bytes=512B
[  111.461965] journalctl[235]: ‌Failed to execute varlink call: Timer expired‌
```

Ignoring the fact, that we do
```bash
+ mkdir -p /etc/systemd/system/systemd-journal-flush.service.d
+ printf '[Service]\nTimeoutSec=180s\n'
```

in the ASan wrapper, this timeout seems to be shorter that the default one (90s).

mrc0mmand commented:
Ah, noticed this warning in the debug logs, maybe that's the culprit here :-)

```
[   15.138981] systemd[1]: /lib/systemd/system/systemd-journal-flush.service:9: Unknown lvalue 'Environment' in section 'Unit', ignoring
```

yuwata commented:
Ouch. Fixed.

mrc0mmand commented:
Thank yout! Now it looks pretty stable, once again:

```
Iteration #11
+ (( i++ ))
+ (( i < 30 ))
+ echo 'Iteration #12'
+ make -C test/TEST-01-BASIC clean setup run QEMU_TIMEOUT=900 TEST_NO_NSPAWN=yes
Iteration #12
##[error]The operation was canceled.
##[Error 1]
The job running on agent Hosted Ubuntu 1604 3 has exceeded the maximum execution time of 60. For more information, see https://go.microsoft.com/fwlink/?linkid=2077134 
```

mrc0mmand commented:
@evverx if the @yuwata's solution makes sense to you, I'll go ahead and drop my debug commits to make the PR merge-able.

mrc0mmand commented:
I dropped the debug-only commits, so this PR can be properly reviewed, PTAL.

yuwata commented:
Thanks.

evverx commented:
The PR looks good to me and more importantly it makes `journalctl` work but if I didn't know why it's written this way it would look weird to me, to be honest. In theory it should be possible to hide that stuff by bumping up the timeout somewhere in `journalctl` if `systemd` is built with ASan (which isn't much better (or probably even worse) in terms of the number of kludges :-). I'd wait for someone else to take a fresh look and decide what to do about it.

evverx commented:
Looks like the commit where `Environment` was moved to the "service" section vanished.

yuwata commented:
> Looks like the commit where `Environment` was moved to the "service" section vanished.

Ouch.

mrc0mmand commented:
> > Looks like the commit where `Environment` was moved to the "service" section vanished.
> 
> Ouch. Fixed.

Ugh, sorry about that, I apparently forgot to pull before doing a force push.

mrc0mmand commented:
Also, I'm not sure if the UI is borked right now, but I still see the `Environment` directive under a `[Unit]` section in the last commit.

yuwata commented:
Hmm? It seems I forgot to push the fix...

yuwata commented:
> The PR looks good to me and more importantly it makes `journalctl` work but if I didn't know why it's written this way it would look weird to me, to be honest. In theory it should be possible to hide that stuff by bumping up the timeout somewhere in `journalctl` if `systemd` is built with ASan (which isn't much better (or probably even worse) in terms of the number of kludges :-). I'd wait for someone else to take a fresh look and decide what to do about it.

`journalctl --smart-relinquish-var` may be run by user, not from service unit. In that case, varlink timeout will be used. Of course, adding `--timeout` option rather than environment variable is also OK for me. Which is preferable?

cc @keszybz.

yuwata commented on units/systemd-journal-flush.service.in:
```diff
@@ -21,4 +21,5 @@ ExecStart=@rootbindir@/journalctl --flush
 Type=oneshot
 RemainAfterExit=yes
+Environment=JOURNALCTL_TIMEOUT=infinity
 TimeoutSec=90s
```
OK.

yuwata commented:
@keszybz Thank you for the comment. I force-pushed a new version. Now the timeout in varlink is simply disabled. PTAL.

keszybz commented:
LGTM!
